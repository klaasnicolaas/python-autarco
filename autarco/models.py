"""Models for Autarco."""
from __future__ import annotations

from dataclasses import dataclass
from typing import Any


@dataclass
class Inverter:
    """Object representing an Inverter model response from the API."""

    serial_number: str
    out_ac_power: int
    out_ac_energy_total: int
    grid_turned_off: bool
    health: str

    @classmethod
    def from_json(cls, data: dict[str, Any]) -> Inverter:
        """Create an Inverter object from a JSON response.

        Args:
            data: JSON response from the API.

        Returns:
            An Inverter object.
        """
        data = data[1]
        return cls(
            serial_number=data.get("sn"),
            out_ac_power=data.get("out_ac_power"),
            out_ac_energy_total=data.get("out_ac_energy_total"),
            grid_turned_off=data.get("grid_turned_off"),
            health=data.get("health"),
        )


@dataclass
class Solar:
    """Object representing a Solar model response from the API."""

    power_production: int
    energy_production_today: int
    energy_production_month: int
    energy_production_total: int

    @staticmethod
    def from_json(data: dict[str, Any]) -> Solar:
        """Create an Solar object from a JSON response.

        Args:
            data: JSON response from the API.

        Returns:
            An Solar object.
        """
        data = data["stats"]["kpis"]
        return Solar(
            power_production=data.get("current_production"),
            energy_production_today=data.get("output_today"),
            energy_production_month=data.get("output_month"),
            energy_production_total=data.get("output_to_date"),
        )


@dataclass
class Account:
    """Object representing an Account model response from the API."""

    public_key: str
    name: str
    city: str
    state: str
    country: str
    timezone: str

    @staticmethod
    def from_json(data: dict[str, Any]) -> Account:
        """Create an Account object from a JSON response.

        Args:
            data: JSON response from the API.

        Returns:
            An Account object.
        """
        data = data["site"]
        address = data["address"]
        return Account(
            public_key=data.get("public_key"),
            name=data.get("name"),
            city=address.get("city"),
            state=address.get("state"),
            country=address.get("country"),
            timezone=data.get("timezone"),
        )
