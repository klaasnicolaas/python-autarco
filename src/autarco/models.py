"""Models for Autarco."""
from __future__ import annotations

from dataclasses import dataclass, field
from datetime import date
from typing import Any

from mashumaro import DataClassDictMixin, field_options
from mashumaro.mixins.orjson import DataClassORJSONMixin


@dataclass
class PowerResponse(DataClassORJSONMixin):
    """Object representing an Inverter model response from the API."""

    inverters: dict[str, Inverter]
    stats: dict[str, dict[str, int]]


@dataclass
class EnergyResponse(DataClassORJSONMixin):
    """Object representing an Inverter model response from the API."""

    stats: dict[str, dict[str, Any]]


@dataclass
class Inverter(DataClassORJSONMixin):
    """Object representing an Inverter model response from the API."""

    serial_number: str = field(metadata=field_options(alias="sn"))
    out_ac_power: int
    out_ac_energy_total: int
    grid_turned_off: bool
    health: str


@dataclass
class Solar(DataClassDictMixin):
    """Object representing a Solar model response from the API."""

    power_production: int = field(metadata=field_options(alias="pv_now"))
    energy_production_today: int = field(metadata=field_options(alias="pv_today"))
    energy_production_month: int = field(metadata=field_options(alias="pv_month"))
    energy_production_total: int = field(metadata=field_options(alias="pv_to_date"))


@dataclass
class Account(DataClassORJSONMixin):
    """Object representing an Account model response from the API."""

    public_key: str
    name: str
    address: Address

    timezone: str
    dt_created: date = field(
        metadata=field_options(
            alias="dt_created", deserialize=lambda x: date.fromisoformat(x)
        )
    )
    has_consumption_meter: bool
    has_battery: bool


@dataclass
class Address(DataClassORJSONMixin):
    """Object representing an Address model response from the API."""

    street: str = field(metadata=field_options(alias="address_line_1"))
    zip_code: str = field(metadata=field_options(alias="postcode"))
    city: str
    country: str
