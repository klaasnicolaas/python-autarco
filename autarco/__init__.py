"""Asynchronous Python client for the Autarco API."""

from .autarco import (
    Autarco,
    AutarcoAuthenticationError,
    AutarcoConnectionError,
    AutarcoConnectionTimeoutError,
    AutarcoError,
)
from .models import Account, Inverter, Solar

__all__ = [
    "Autarco",
    "AutarcoConnectionError",
    "AutarcoConnectionTimeoutError",
    "AutarcoAuthenticationError",
    "AutarcoError",
    "Inverter",
    "Solar",
    "Account",
]
