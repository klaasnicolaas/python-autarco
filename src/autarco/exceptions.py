"""Exceptions for Autarco."""


class AutarcoError(Exception):
    """Generic Autarco exception."""


class AutarcoConnectionError(AutarcoError):
    """Autarco connection exception."""


class AutarcoAuthenticationError(AutarcoError):
    """Autarco Authentication exception."""


class AutarcoConnectionTimeoutError(AutarcoError):
    """Autarco connection timeout exception."""
