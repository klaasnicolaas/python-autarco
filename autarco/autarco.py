"""Asynchronous Python client for Autarco API."""
from __future__ import annotations

import asyncio
import socket
from collections.abc import Mapping
from dataclasses import dataclass
from importlib import metadata
from typing import Any

import aiohttp
import async_timeout
from aiohttp.client import ClientError, ClientResponseError, ClientSession
from aiohttp.hdrs import METH_GET
from yarl import URL

from .exceptions import (
    AutarcoAuthenticationError,
    AutarcoConnectionError,
    AutarcoConnectionTimeoutError,
    AutarcoError,
)
from .models import Account, Inverter, Solar


@dataclass
class Autarco:
    """Main class for handling connections to Autarco."""

    public_key: str
    username: str
    password: str

    request_timeout: float = 10.0
    session: ClientSession | None = None

    _close_session: bool = False

    async def _request(
        self,
        uri: str,
        *,
        method: str = METH_GET,
        params: Mapping[str, str] | None = None,
    ) -> dict[str, Any]:
        """Handle a request to the Autarco API.

        A generic method for sending/handling HTTP requests done against
        the Autarco API.

        Args:
            uri: Request URI, without '/', for example, 'status'.
            method: HTTP method to use.
            params: Dictionary of parameters to send to the Autarco API.

        Returns:
            The response data from the Autarco API.

        Raises:
            AutarcoAuthenticationError: If the username or password is invalid.
            AutarcoConnectionError: An error occurred while communicating
                with the Autarco API.
            AutarcoConnectionTimeoutError: A timeout occurred while communicating
                with the Autarco API.
            AutarcoError: Received an unexpected response from the
                Autarco API.
        """
        version = metadata.version(__package__)
        url = URL.build(scheme="https", host="my.autarco.com", path="/api/site/").join(
            URL(uri)
        )

        headers = {
            "Accept": "application/json",
            "User-Agent": f"PythonAutarco/{version}",
        }

        if self.session is None:
            self.session = ClientSession()
            self._close_session = True

        # Set basic auth credentials.
        auth = aiohttp.BasicAuth(self.username, self.password)

        try:
            async with async_timeout.timeout(self.request_timeout):
                response = await self.session.request(
                    method,
                    url,
                    auth=auth,
                    headers=headers,
                    params=params,
                    ssl=True,
                )
                response.raise_for_status()
        except asyncio.TimeoutError as exception:
            raise AutarcoConnectionTimeoutError(
                "Timeout occurred while connecting to Autarco API"
            ) from exception
        except ClientResponseError as exception:
            if exception.status == 401:
                raise AutarcoAuthenticationError(
                    "Authentication to the Autarco API failed"
                ) from exception
            raise AutarcoConnectionError(
                "Error occurred while connecting to the Autarco API"
            ) from exception
        except (ClientError, socket.gaierror) as exception:
            raise AutarcoConnectionError(
                "Error occurred while communicating with the Autarco API"
            ) from exception

        content_type = response.headers.get("Content-Type", "")
        if "application/json" not in content_type:
            text = await response.text()
            raise AutarcoError(
                "Unexpected response from the Autarco API",
                {"Content-Type": content_type, "response": text},
            )

        return await response.json()

    async def all_inverters(self) -> Inverter:
        """Get a list of all used inverters.

        Returns:
            A list of Inverter objects.
        """
        results = []

        data = await self._request(f"{self.public_key}/power")
        for item in data["inverters"].items():
            results.append(Inverter.from_json(item))
        return results

    async def solar(self) -> Solar:
        """Get.

        Returns:
            An Solar object.
        """

        data = await self._request(f"{self.public_key}/")
        return Solar.from_json(data)

    async def account(self) -> Account:
        """Get.

        Returns:
            An Account object.
        """

        data = await self._request(f"{self.public_key}/")
        return Account.from_json(data)

    async def close(self) -> None:
        """Close open client session."""
        if self.session and self._close_session:
            await self.session.close()

    async def __aenter__(self) -> Autarco:
        """Async enter.

        Returns:
            The Autarco object.
        """
        return self

    async def __aexit__(self, *_exc_info) -> None:
        """Async exit.

        Args:
            _exc_info: Exec type.
        """
        await self.close()
