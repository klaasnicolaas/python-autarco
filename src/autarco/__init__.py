"""Asynchronous Python client for the Autarco API."""

from .autarco import Autarco
from .exceptions import (
    AutarcoAuthenticationError,
    AutarcoConnectionError,
    AutarcoConnectionTimeoutError,
    AutarcoError,
)
from .models import Account, Inverter, Solar

__all__ = [
    "Autarco",
    "AutarcoConnectionError",
    "AutarcoConnectionTimeoutError",
    "AutarcoAuthenticationError",
    "AutarcoError",
    "Inverter",
    "Solar",
    "Account",
]
