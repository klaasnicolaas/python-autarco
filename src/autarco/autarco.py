"""Asynchronous Python client for Autarco API."""

from __future__ import annotations

import asyncio
import json
import socket
from dataclasses import dataclass
from importlib import metadata
from typing import Any, Self

from aiohttp import BasicAuth, ClientError, ClientResponseError, ClientSession
from aiohttp.hdrs import METH_GET
from yarl import URL

from .exceptions import (
    AutarcoAuthenticationError,
    AutarcoConnectionError,
    AutarcoError,
)
from .models import (
    AccountResponse,
    AccountSite,
    Battery,
    EnergyResponse,
    Inverter,
    PowerResponse,
    Site,
    Solar,
    Stats,
)

VERSION = metadata.version(__package__)


@dataclass
class Autarco:
    """Main class for handling connections to Autarco."""

    email: str
    password: str

    request_timeout: float = 15.0
    session: ClientSession | None = None

    _close_session: bool = False

    async def _request(
        self,
        uri: str,
        *,
        method: str = METH_GET,
        params: dict[str, Any] | None = None,
    ) -> str:
        """Handle a request to the Autarco API.

        A generic method for sending/handling HTTP requests done against
        the Autarco API.

        Args:
        ----
            uri: Request URI, without '/', for example, 'status'.
            method: HTTP method to use.
            params: Query parameters to send with the request.

        Returns:
        -------
            The response data from the Autarco API.

        Raises:
        ------
            AutarcoAuthenticationError: If the email or password is invalid.
            AutarcoConnectionError: An error occurred while communicating
                with the Autarco API.
            AutarcoConnectionTimeoutError: A timeout occurred while communicating
                with the Autarco API.
            AutarcoError: Received an unexpected response from the
                Autarco API.

        """
        url = URL.build(
            scheme="https",
            host="my.autarco.com",
            path="/api/site/",
        ).join(URL(uri))

        headers = {
            "Accept": "application/json",
            "User-Agent": f"PythonAutarco/{VERSION}",
        }

        if self.session is None:
            self.session = ClientSession()
            self._close_session = True

        # Set basic auth credentials.
        auth = BasicAuth(self.email, self.password)

        try:
            async with asyncio.timeout(self.request_timeout):
                response = await self.session.request(
                    method,
                    url,
                    auth=auth,
                    headers=headers,
                    params=params,
                    ssl=True,
                )
                response.raise_for_status()
        except TimeoutError as exception:
            msg = "Timeout occurred while connecting to Autarco API"
            raise AutarcoConnectionError(msg) from exception
        except ClientResponseError as exception:
            if exception.status == 401:
                msg = "Authentication to the Autarco API failed"
                raise AutarcoAuthenticationError(msg) from exception
            msg = "Error occurred while communicating with the Autarco API"
            raise AutarcoConnectionError(msg) from exception
        except (ClientError, socket.gaierror) as exception:
            msg = "Error occurred while communicating with the Autarco API"
            raise AutarcoConnectionError(msg) from exception

        content_type = response.headers.get("Content-Type", "")
        text = await response.text()
        if "application/json" not in content_type:
            msg = "Unexpected response from the Autarco API"
            raise AutarcoError(
                msg,
                {"Content-Type": content_type, "response": text},
            )

        return text

    async def _get_combined_data(self, public_key: str) -> dict[str, Any]:
        """Get a combined dictionary with power and energy data from a site.

        Args:
        ----
            public_key: The public key from your site.

        Returns:
        -------
            A dictionary with combined power and energy data.

        """
        power_response = await self._request(f"{public_key}/kpis/power")
        energy_response = await self._request(f"{public_key}/kpis/energy")
        return {**json.loads(power_response), **json.loads(energy_response)}

    async def get_account(self) -> list[AccountSite]:
        """Get account with list of sites.

        Returns
        -------
            A list of Site objects.

        """
        response = await self._request("")
        return AccountResponse.from_json(response).sites

    async def get_inverters(self, public_key: str) -> dict[str, Inverter]:
        """Get a list of all used inverters.

        Args:
        ----
            public_key: The public key from the specific site.

        Returns:
        -------
            A list of Inverter objects.

        """
        response = await self._request(f"{public_key}/power")
        return PowerResponse.from_json(response).inverters

    async def get_power_statistics(
        self, public_key: str, query_range: str = "day"
    ) -> Stats:
        """Get the statistics of the inverters.

        Args:
        ----
            public_key: The public key from the specific site.
            query_range: The range of time to get the statistics for.

        Returns:
        -------
            A list of Inverter objects.

        """
        response = await self._request(f"{public_key}/power", params={"r": query_range})
        return PowerResponse.from_json(response).stats

    async def get_energy_statistics(
        self, public_key: str, query_range: str = "month"
    ) -> Stats:
        """Get the statistics of the inverters.

        Args:
        ----
            public_key: The public key from the specific site.
            query_range: The range of time to get the statistics for.

        Returns:
        -------
            A list of Inverter objects.

        """
        response = await self._request(
            f"{public_key}/energy", params={"r": query_range}
        )
        return EnergyResponse.from_json(response).stats

    async def get_solar(self, public_key: str) -> Solar:
        """Get information about the solar production from a site.

        Args:
        ----
            public_key: The public key from your site.

        Returns:
        -------
            An Solar object.

        """
        combined_data = await self._get_combined_data(public_key)
        return Solar.from_dict(combined_data)

    async def get_site(self, public_key: str) -> Site:
        """Get information about your system site.

        Args:
        ----
            public_key: The public key from your site.

        Returns:
        -------
            An Site object.

        """
        response = await self._request(f"{public_key}/")
        return Site.from_json(response)

    async def get_battery(self, public_key: str) -> Battery:
        """Get information about the battery from a site.

        Args:
        ----
            public_key: The public key from your site.

        Returns:
        -------
            An Battery object.

        """
        combined_data = await self._get_combined_data(public_key)
        return Battery.from_dict(combined_data)

    async def close(self) -> None:
        """Close open client session."""
        if self.session and self._close_session:
            await self.session.close()

    async def __aenter__(self) -> Self:
        """Async enter.

        Returns
        -------
            The Autarco object.

        """
        return self

    async def __aexit__(self, *_exc_info: object) -> None:
        """Async exit.

        Args:
        ----
            _exc_info: Exec type.

        """
        await self.close()
