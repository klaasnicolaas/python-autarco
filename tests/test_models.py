"""Test the Autarco models."""
from aiohttp import ClientSession
from aresponses import ResponsesMockServer
from syrupy.assertion import SnapshotAssertion

from autarco import Account, Autarco, Inverter, Solar

from . import load_fixtures


async def test_get_inverters(
    aresponses: ResponsesMockServer,
    snapshot: SnapshotAssertion,
    autarco_client: Autarco,
) -> None:
    """Test request from a Autarco API - Inverter object."""
    aresponses.add(
        "my.autarco.com",
        "/api/site/fake_key/power",
        "GET",
        aresponses.Response(
            text=load_fixtures("power.json"),
            status=200,
            headers={"Content-Type": "application/json; charset=utf-8"},
        ),
    )
    inverters: dict[str, Inverter] = await autarco_client.get_inverters(
        public_key="fake_key"
    )
    assert inverters == snapshot


async def test_get_solar(
    aresponses: ResponsesMockServer,
    snapshot: SnapshotAssertion,
    autarco_client: Autarco,
) -> None:
    """Test request from a Autarco API - Solar object."""
    # Energy response
    aresponses.add(
        "my.autarco.com",
        "/api/site/fake_key/energy",
        "GET",
        aresponses.Response(
            text=load_fixtures("energy.json"),
            status=200,
            headers={"Content-Type": "application/json; charset=utf-8"},
        ),
    )
    # Power response
    aresponses.add(
        "my.autarco.com",
        "/api/site/fake_key/power",
        "GET",
        aresponses.Response(
            text=load_fixtures("power.json"),
            status=200,
            headers={"Content-Type": "application/json; charset=utf-8"},
        ),
    )
    solar: Solar = await autarco_client.get_solar(public_key="fake_key")
    assert solar == snapshot


async def test_get_account(
    aresponses: ResponsesMockServer,
    snapshot: SnapshotAssertion,
    autarco_client: Autarco,
) -> None:
    """Test request from a Autarco API - Account object."""
    aresponses.add(
        "my.autarco.com",
        "/api/site/fake_key/",
        "GET",
        aresponses.Response(
            text=load_fixtures("account.json"),
            status=200,
            headers={"Content-Type": "application/json; charset=utf-8"},
        ),
    )
    account: Account = await autarco_client.get_account(public_key="fake_key")
    assert account == snapshot


async def test_get_public_key(
    aresponses: ResponsesMockServer,
    snapshot: SnapshotAssertion,
    autarco_client: Autarco,
) -> None:
    """Test request from a Autarco API - get_public_key."""
    aresponses.add(
        "my.autarco.com",
        "/api/site/",
        "GET",
        aresponses.Response(
            text=load_fixtures("public_key.json"),
            status=200,
            headers={"Content-Type": "application/json; charset=utf-8"},
        ),
    )
    public_key = await autarco_client.get_public_key()
    assert public_key == snapshot
    assert public_key == "sd6fv516"
