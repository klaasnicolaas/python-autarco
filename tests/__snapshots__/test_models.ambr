# serializer version: 1
# name: test_get_account
  Account(public_key='sd6fv516', name='My Autarco solar installation', address=Address(street='Streetname 00', zip_code='1111 AA', city='Valkenswaard', country='Nederland'), timezone='Europe/Amsterdam', dt_created=datetime.date(2023, 5, 15), has_consumption_meter=False, has_battery=False)
# ---
# name: test_get_inverters
  dict({
    '123456789876': Inverter(serial_number='123456789876', out_ac_power=100, out_ac_energy_total=6605, grid_turned_off=False, health='OK'),
    '987654321234': Inverter(serial_number='987654321234', out_ac_power=100, out_ac_energy_total=3607, grid_turned_off=False, health='OK'),
  })
# ---
# name: test_get_public_key
  'sd6fv516'
# ---
# name: test_get_solar
  Solar(power_production=200, energy_production_today=4, energy_production_month=58, energy_production_total=10379)
# ---
